### PR TITLE
Connection pool

### PR DESCRIPTION
**Pull request by nicklester**
First pass at this, Eli. Appreciate some thoughts.

### CODE REVIEW

**Comment by AnnaGrebnev** [comment]
@e-orz and @nicklester what's issue id it's related to?

**Comment by e-orz** [comment]
to allow shared connection pool for all the agents of STP and reduce the load on the server.

**Comment by nicklester** [comment]
There is no raised issue, but I'll create one shortly.

File: server/cmwell-data-tools/src/main/resources/reference.conf [hunk 1]
```diff
@@ -1,17 +1,21 @@
-akka {
-  log-dead-letters = 2
-  log-dead-letters-during-shutdown = on
-  #  log-config-on-start = on
-  #  jvm-exit-on-fatal-error = on
+data-tools {
+  akka {
+    log-dead-letters = 2
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/resources/reference.conf]: IMHO, the akka logging parameters aren't data-tools specific. You can leave it to the default ones configured by the main application.

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferFillerActor.scala [hunk 2]
```diff
@@ -91,7 +91,7 @@ class BufferFillerActor(threshold: Int,
   private val HttpAddress(protocol, host, port, _) =
     ArgsManipulations.extractBaseUrl(baseUrl)
   private val conn =
-    HttpConnections.newHostConnectionPool[Option[_]](host, port, protocol)
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferFillerActor.scala]: I think that also in here there should be the `config` parameter.

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/HttpConnections.scala [hunk 3]
```diff
@@ -19,60 +19,29 @@ import akka.http.scaladsl._
 import akka.http.scaladsl.settings.{ClientConnectionSettings, ConnectionPoolSettings}
 import akka.stream.Materializer
 import cmwell.tools.data.utils.logging.LabelId
-import com.typesafe.config.ConfigFactory
 
 object HttpConnections extends DataToolsConfig {
 
   def outgoingConnection(host: String, port: Int, protocol: String = "http")(implicit system: ActorSystem,
-                                                                             label: Option[LabelId] = None) = {
-    val userAgent = label.fold(s"cmwell-data-tools using akka-http/${config.getString("akka.version")}")(
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/HttpConnections.scala]: It's my mistake I told you to remove it. It has the label that is injected to the header.
It's up to you whether to leave the label-in-header as a debugging aid or not.

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/HttpConnections.scala [hunk 4]
```diff
@@ -19,60 +19,29 @@ import akka.http.scaladsl._
...
-      case _       => Http().outgoingConnection(host, port, settings = settings)
-    }
+    case "https" => Http().outgoingConnectionHttps(host, port, settings = ClientConnectionSettings(config))
+    case _       => Http().outgoingConnection(host, port, settings = ClientConnectionSettings(config))
   }
 
   def newHostConnectionPool[T](host: String, port: Int, protocol: String = "http")(implicit system: ActorSystem,
                                                                                    mat: Materializer,
-                                                                                   label: Option[LabelId] = None) = {
-    val userAgent = label.fold(s"cmwell-data-tools using akka-http/${config.getString("akka.version")}")(
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/HttpConnections.scala]: same here

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/HttpConnections.scala [hunk 5]
```diff
@@ -19,60 +19,29 @@ import akka.http.scaladsl._
...
-      case _       => Http().newHostConnectionPool[T](host, port, settings = settings)
-    }
+    case "https" => Http().newHostConnectionPoolHttps[T](host, port, settings = ConnectionPoolSettings(config))
+    case _       => Http().newHostConnectionPool[T](host, port, settings = ConnectionPoolSettings(config))
   }
 
   def cachedHostConnectionPool[T](host: String, port: Int, protocol: String = "http")(implicit system: ActorSystem,
                                                                                       mat: Materializer,
-                                                                                      label: Option[LabelId] = None) = {
-    val userAgent = label.fold(s"cmwell-data-tools using akka-http/${config.getString("akka.version")}")(
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/HttpConnections.scala]: same here

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala [hunk 6]
```diff
@@ -258,7 +258,7 @@ object Retry extends DataToolsLogging with DataToolsConfig {
 //    val conn = Http().newHostConnectionPool[State](host = baseUrl, port = port) // http connection flow
     val HttpAddress(protocol, host, port, uriPrefix) =
       ArgsManipulations.extractBaseUrl(baseUrl)
-    val conn = HttpConnections.newHostConnectionPool[State](
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: There shouldn't be a config parameter here?

File: server/cmwell-dc/src/pack/application.conf [hunk 7]
```diff
@@ -27,6 +27,19 @@ akka {
   }
 }
 
+data-tools {
+  akka {
+    http.host-connection-pool {
+      max-connections = 20
+      max-open-requests = 128
+      pipelining-limit = 1
```
Comment (e-orz) [inline server/cmwell-dc/src/pack/application.conf]: This is the default, why changing it?

**Pull request merged**